### PR TITLE
v2ray-geoip: init at 202112090029

### PR DESCRIPTION
**Pull request by NickCao**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Vonfry** [comment]
What is the difference from geoip in [v2ray](https://github.com/NixOS/nixpkgs/blob/0f0ffc5e867cbd93134c13396174f7cc6ebe10a9/pkgs/tools/networking/v2ray/default.nix#L17-L36)? 

Furthermore, should we replace the one in v2ray with this?

**Comment by NickCao** [comment]
It's the same as the one bundled in v2ray. I think it is more appropriate to replace that one with this, and would cover that in a separate pr.

**Review by VergeDX — APPROVED** [review]

**Comment by kidonng** [comment]
> The reason for a separate package is that those assets are usable outside the context of v2ray (as in other proxy clients), and the current expressions makes it hard to do so.

Agreed. Without this package a hack is needed to access those assets:

```nix
let assets = (pkgs.runCommandLocal "v2ray-assets" {} ''
      mkdir -p $out/share
      ln -s `[[ "${pkgs.v2ray.drvAttrs.buildCommand}" =~ V2RAY_LOCATION_ASSET( ([[:alnum:]]|[/-])+) ]] && echo ''${BASH_REMATCH[1]}` $out/share/v2ray
'')
```

**Review by kidonng — APPROVED** [review]

File: pkgs/data/misc/v2ray-geoip/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{ lib, stdenv, fetchFromGitHub }:
+
+stdenv.mkDerivation rec {
```
Comment (sikmir) [inline pkgs/data/misc/v2ray-geoip/default.nix:3]: Could be simpler:
```
let
  version = "202111250029";
in
fetchurl {
  name = "v2ray-geoip-${version}";
  url = ""https://github.com/v2fly/geoip/releases/download/${version}/geoip.dat";
  sha256 = ...

  downloadToTemp = true;
  recursiveHash = true;

  postFetch = "install -Dm0644 $downloadedFile $out/share/v2ray/geoip.dat";
  meta = ...
}
```

Comment (NickCao) [reply pkgs/data/misc/v2ray-geoip/default.nix:3]: However this approach makes it hard to do things like `nix-prefetch-url`

Comment (sikmir) [reply pkgs/data/misc/v2ray-geoip/default.nix:3]: Why? `nix-prefetch-url https://github.com/v2fly/geoip/releases/download/202111250029/geoip.dat`

Comment (NickCao) [reply pkgs/data/misc/v2ray-geoip/default.nix:3]: `recursiveHash = true;` is for the output hash, not the input.

**Pull request merged**